pkg/subnet/local.go

```diff
@@ -123,54 +122,171 @@ func (d *Deployer) doDeploy(chain string, chainGenesis string) error {
+		toInstallVMIDs[vmID] = struct{}{}
+	}
+	toInstallVMIDs[toDeployVMID.String()] = struct{}{}
 	binDir := filepath.Join(d.app.GetBaseDir(), constants.AvalancheCliBinDir)
-	if err := d.binaryDownloader.Download(vmID, pluginDir, binDir); err != nil {
+	if err := d.binaryDownloader.Download(toInstallVMIDs, pluginDir, binDir); err != nil {
 		return err
 	}
 
-	ctx := binutils.GetAsyncContext()
+	ux.Logger.PrintToUser("VM ready.")
+
+	// do start/restart as needed
+	if needsStart {
+		ux.Logger.PrintToUser("Starting network...")
+		loadSnapshotOpts := []client.OpOption{
+			client.WithPluginDir(pluginDir),
+			client.WithExecPath(avalancheGoBinPath),
+			client.WithRootDataDir(runDir),
+		}
+		loadSnapshotsInfo, err := cli.LoadSnapshot(
+			ctx,
+			constants.DefaultSnapshotName,
+			loadSnapshotOpts...,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to start network :%s", err)
+		}
+		d.app.Log.Debug(loadSnapshotsInfo.String())
+	} else {
+		// we need to restart the network if the VM was not deployed previously
+		needsRestart := false
+		_, ok := deployedVMIDs[toDeployVMID.String()]
+		if !ok {
+			needsRestart = true
+		}
+
+		if needsRestart {
+			// make snapshot of current state and reload it again
+			tmpSnapshotName := fmt.Sprintf("restart-tmp-%d", time.Now().Unix())
+			ux.Logger.PrintToUser("Restarting network...")
+			_, err := cli.SaveSnapshot(
+				ctx,
+				tmpSnapshotName,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to save snapshot :%s", err)
+			}
+			loadSnapshotOpts := []client.OpOption{
+				client.WithPluginDir(pluginDir),
+				client.WithExecPath(avalancheGoBinPath),
+				client.WithRootDataDir(runDir),
+			}
+			_, err = cli.LoadSnapshot(
+				ctx,
+				tmpSnapshotName,
+				loadSnapshotOpts...,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to load snapshot :%s", err)
+			}
+			_, err = cli.RemoveSnapshot(
+				ctx,
+				tmpSnapshotName,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to remove snapshot :%s", err)
+			}
+		}
+	}
+
+	clusterInfo, err = d.WaitForHealthy(ctx, cli, d.healthCheckInterval)
+	if err != nil {
+		return fmt.Errorf("failed to query network health: %s", err)
+	}
+	subnetIDs := clusterInfo.Subnets
+	numBlockchains := len(clusterInfo.CustomVms)
+
+	// in order to make subnet deploy faster, a set of validated subnet IDs is preloaded
+	// in the bootstrap snapshot
+	// we select one to be used for creating the next blockchain, for that we use the
+	// number of currently created blockchains as the index to select the next subnet ID,
+	// so we get incremental selection
+	sort.Strings(subnetIDs)
+	subnetID := ""
+	// in unit tests, there are no preloaded subnets IDs
+	// also, for the case the network does not contain subnet IDs, empty subnet ID
+	// will make ANR to create one when creating the blockchain
+	if len(subnetIDs) > 0 {
+		subnetID = subnetIDs[numBlockchains%len(subnetIDs)]
+	}
+
+	blockchainSpecs := []*rpcpb.BlockchainSpec{
```

Comment:
Add documentation here why you need to do this now against how it was before. 

Please also add some general flow description to the start of the function about what is happening during a complete deploy, there's a lot going on and I needed quite some time to understand all pieces. It is only clear to us because of the way we changed the ANR, otherwise I think a reader will have a hard time.